File: examples/python-agent/sqlrite_agent/sqlutil.py
"""SQL-literal helpers.

The SQLRite Python SDK does not support prepared-statement parameter
binding yet (deferred to engine Phase 5a.2 — see ``sdk/python/README.md``).
Every value must therefore be inlined into the SQL string. These helpers
keep that inlining safe and consistent across the agent.
"""

from __future__ import annotations

from typing import Iterable


def q(value: object) -> str:
    """Render ``value`` as a SQLRite literal safe for direct inlining.

    Strings get single-quote-escaped (``'`` → ``''``). Numbers and bools
    pass through. ``None`` becomes ``NULL``. Lists / tuples of floats
    become the ``[x, y, z]`` syntax SQLRite uses for VECTOR literals.
    """
    if value is None:
        return "NULL"
    if isinstance(value, bool):
        # Cover bool before int — bool is a subclass of int.
        return "1" if value else "0"
    if isinstance(value, (int, float)):
        return repr(value)
    if isinstance(value, (list, tuple)):
        return vec_literal(value)
    if isinstance(value, str):
        escaped = value.replace("'", "''")
        return f"'{escaped}'"
    raise TypeError(f"cannot inline {type(value).__name__} into SQL")


def vec_literal(vec: Iterable[float]) -> str:
    """Render a vector as the SQLRite ``[v1, v2, ...]`` literal."""
    parts = ", ".join(f"{float(x):.6f}" for x in vec)
    return f"[{parts}]"

File: examples/nodejs-notes/src/sqlutil.mjs
// Tiny SQL-literal helpers — the SQLRite engine doesn't support
// `?`-style parameter binding yet (Phase 5a.2 follow-up), so every
// caller must inline values as SQL literals. This module is the
// single place that does that safely.
//
// Mirrors the shape of `sqlrite_agent.sqlutil` in the Python example.

/**
 * Quote a JavaScript value as a SQL literal.
 *
 * - string  → `'escaped'` (single quotes doubled per the SQL standard)
 * - number  → integer or `Number.prototype.toString()` for finite floats
 * - boolean → `TRUE` / `FALSE`
 * - null/undefined → `NULL`
 * - number[] → `[v1, v2, ...]` — the engine's vector literal syntax
 *
 * Anything else throws — refuse to silently `String()` an object.
 *
 * @param {unknown} value
 * @returns {string}
 */
export function q(value) {
  if (value === null || value === undefined) return 'NULL';
  if (typeof value === 'string') return `'${value.replaceAll("'", "''")}'`;
  if (typeof value === 'number') {
    if (!Number.isFinite(value)) {
      throw new TypeError(`q(): non-finite number ${value}`);
    }
    return Number.isInteger(value) ? String(value) : value.toString();
  }
  if (typeof value === 'bigint') return value.toString();
  if (typeof value === 'boolean') return value ? 'TRUE' : 'FALSE';
  if (Array.isArray(value)) {
    // Vector literal — every element must be finite numeric.
    const parts = value.map((v, i) => {
      if (typeof v !== 'number' || !Number.isFinite(v)) {
        throw new TypeError(`q(): vector element ${i} is not a finite number (got ${v})`);
      }
      // toString() emits the shortest round-trippable form; the
      // engine's parser accepts both fixed-point and exponential.
      return v.toString();
    });
    return `[${parts.join(', ')}]`;
  }
  throw new TypeError(`q(): unsupported value type ${typeof value}`);
}

/**
 * Validate a SQL identifier (table / column / index name) against the
 * unquoted-identifier subset the engine accepts. Throws if invalid.
 *
 * Use this for ANY identifier that ultimately gets inlined into SQL —
 * callers shouldn't have to guess what's safe.
 *
 * @param {string} name
 * @returns {string} the same name (for chaining)
 */
export function ident(name) {
  if (typeof name !== 'string' || !/^[A-Za-z_][A-Za-z0-9_]*$/.test(name)) {
    throw new TypeError(`ident(): invalid SQL identifier ${JSON.stringify(name)}`);
  }
  return name;
}
